packages/xod-client/src/hinting/validation.internal.js

```diff
@@ -0,0 +1,471 @@
+  (project, allDeducedPinTypes, changedPatchPath, prevErrors) =>
+    R.compose(
+      validatePatchesGenerally(project, allDeducedPinTypes, prevErrors),
+      filterPatchAndDependentPatchesByPatchPath(changedPatchPath),
+      XP.listPatches
+    )(project)
+);
+
+/**
+ * Validates only changed Patch, if Action payload has a `patchPath` property
+ * and sets MERGE policy. If there is no PatchPath it will check all Patches with
+ * OVERWRITE policy.
+ *
+ * But it's an utility function for short-circuits and it should not been used
+ * for Actions, that does not contain `PatchPath`.
+ *
+ * validateChangedPatch ::
+ *    [NodeValidateFn] ->
+ *    [PinValidateFn] ->
+ *    [LinkValidateFn] ->
+ *    Action ->
+ *    Project ->
+ *    Map PatchPath DeducedPinTypes ->
+ *    Map PatchPath PatchErrors ->
+ *    Map PatchPath PatchErrors
+ */
+export const validateChangedPatch = R.curry(
+  (
+    nodeValidators,
+    pinValidators,
+    linkValidators,
+    action,
+    project,
+    allDeducedPinTypes,
+    prevErrors
+  ) => {
+    const validateFn = validatePatches(
+      [R.always({}), ...nodeValidators],
+      [R.always({}), ...pinValidators],
+      [R.always([]), ...linkValidators]
+    );
+
+    return foldMaybeWith(
+      // No PatchPath -> Throw an Error, cause it developers bug
+      () => {
+        throw new Error(
+          `Hinting middleware has a wrong short-circuit validation for action type: ${
+            action.type
+          }.`
+        );
+      },
+      // PatchPath exists -> Validate only changed Patch
+      patchPath =>
+        R.compose(
+          setMergePolicy,
+          validateFn(project, allDeducedPinTypes, prevErrors),
+          R.of,
+          XP.getPatchByPathUnsafe
+        )(patchPath, project),
+      // Try to get PatchPath from Action
+      getActingPatchPath(action)
+    );
+  }
+);
+
+/**
+ * Validates only changed Patch, if Action changed only one Patch.
+ * If it has any errors — it will check then local patches (if changed Patch was
+ * in a local project) or check all patches (if changed Patch was in a library).
+ *
+ * It automatically sets merge policy depending on list of changed Patches.
+ *
+ * It will be called in middleware on each Action, that changed Project, if
+ * there will no short-circuit validation function and validation should not be
+ * skipped for this ActionType.
+ *
+ * defaultValidateFunction ::
+ *    Action ->
+ *    Project ->
+ *    Map PatchPath DeducedPinTypes ->
+ *    Map PatchPath PatchErrors ->
+ *    ErrorsUpdateData
+ */
+export const defaultValidateFunction = (
+  action,
+  project,
+  allDeducedPinTypes,
+  prevErrors
+) =>
+  foldMaybeWith(
+    // No PatchPath -> Validate All
+    () =>
+      R.compose(setOverwritePolicy, validateAllPatches)(
+        project,
+        allDeducedPinTypes,
+        null,
+        prevErrors
+      ),
+    // PatchPath exists -> Validate this Patch first
+    patchPath =>
+      R.compose(
+        R.compose(
```

Comment:
`R.compose(` of `R.compose(` can be replaced with a single `R.compose` which removes one nesting level